File: plugins/plugin-code/rust/src/actions/list_files.rs
use async_trait::async_trait;
use serde_json::Value;

use crate::service::CoderService;
use crate::{Action, ActionExample, ActionResult};

pub struct ListFilesAction;

#[async_trait]
impl Action for ListFilesAction {
    fn name(&self) -> &str {
        "LIST_FILES"
    }

    fn similes(&self) -> Vec<&str> {
        vec!["LS", "LIST_DIR", "LIST_DIRECTORY", "DIR"]
    }

    fn description(&self) -> &str {
        "List files in a directory."
    }

    async fn validate(&self, _message: &Value, _state: &Value) -> bool {
        true
    }

    async fn handler(
        &self,
        message: &Value,
        state: &Value,
        service: Option<&mut CoderService>,
    ) -> ActionResult {
        let Some(svc) = service else {
            return ActionResult {
                success: false,
                text: "Coder service is not available.".to_string(),
                data: None,
                error: Some("missing_service".to_string()),
            };
        };

        let dirpath = state.get("path").and_then(|v| v.as_str()).unwrap_or(".");
        let conv = message
            .get("room_id")
            .and_then(|v| v.as_str())
            .or_else(|| message.get("agent_id").and_then(|v| v.as_str()))
            .unwrap_or("default");

        match svc.list_files(conv, dirpath).await {
            Ok(items) => ActionResult {
                success: true,
                text: if items.is_empty() {
                    "(empty)".to_string()
                } else {
                    items.join("\n")
                },
                data: None,
                error: None,
            },
            Err(err) => ActionResult {
                success: false,
                text: err,
                data: None,
                error: Some("list_failed".to_string()),
            },
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![ActionExample {
            user_message: "list files".to_string(),
            agent_response: "Listing…".to_string(),
        }]
    }
}

File: plugins/plugin-code/typescript/actions/listFiles.ts
import type {
  Action,
  ActionResult,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import type { CoderService } from "../services/coderService";

function getDir(options: HandlerOptions | undefined): string {
  const opt = options as { path?: string } | undefined;
  return opt?.path?.trim() ?? ".";
}

const spec = requireActionSpec("LIST_FILES");

export const listFiles: Action = {
  name: spec.name,
  similes: spec.similes ? [...spec.similes] : [],
  description: spec.description,
  validate: async (runtime: IAgentRuntime): Promise<boolean> =>
    runtime.getService<CoderService>("coder") !== null,
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state: State | undefined,
    options: HandlerOptions | undefined,
    callback?: HandlerCallback,
  ): Promise<ActionResult> => {
    const svc = runtime.getService<CoderService>("coder");
    if (!svc) {
      const msg = "Coder service is not available.";
      if (callback) await callback({ content: { text: msg } });
      return { success: false, text: msg };
    }

    const conv = message.roomId ?? message.agentId ?? runtime.agentId;
    const dir = getDir(options);
    const result = await svc.listFiles(conv, dir);
    if (!result.ok) {
      if (callback) await callback({ content: { text: result.error } });
      return { success: false, text: result.error };
    }

    const out = result.items.length > 0 ? result.items.join("\n") : "(empty)";
    if (callback) await callback({ content: { text: out } });
    return { success: true, text: out };
  },
};
